File: leetcode/binarytree/lowestcommonancestor.go
package binarytree

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	return lcaTraversal(root, p, q)
}

func lcaTraversal(root *TreeNode, p, q *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	if root == p || root == q {
		return root
	}

	left := lcaTraversal(root.Left, p, q)
	right := lcaTraversal(root.Right, p, q)
	if left != nil && right != nil {
		return root
	}

	if left != nil {
		return left
	}
	return right
}

File: Interview/zoom/cpp/lowest_common_ancestor.cpp
#include <unordered_set>
#include <vector>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* lowestCommonAncestor4(TreeNode* root, std::vector<TreeNode*>& nodes) {
    std::unordered_set<TreeNode*> nodeSet;
    for (TreeNode* node : nodes) {
        nodeSet.insert(node);
    }
    
    std::function<TreeNode*(TreeNode*)> dfs = [&](TreeNode* root) -> TreeNode* {
        if (root == nullptr) {
            return nullptr;
        }
        if (nodeSet.find(root) != nodeSet.end()) {
            return root;
        }

        TreeNode* left = dfs(root->left);
        TreeNode* right = dfs(root->right);

        if (left && right) {
            return root;
        }
        return left ? left : right;
    };

    return dfs(root);
}